chunk.go

```diff
@@ -0,0 +1,159 @@
+package iavl
+
+import (
+	"sort"
+)
+
+// Chunk is a list of ordered nodes
+// It can be sorted, merged, exported from a tree and
+// used to generate a new tree
+type Chunk []OrderedNodeData
+
+// OrderedNodeData is the data to recreate a leaf node,
+// along with a SortOrder to define a BFS insertion order
+type OrderedNodeData struct {
+	SortOrder uint64
+	NodeData
+}
+
+// NewOrderedNode creates the data from a leaf node
+func NewOrderedNode(leaf *Node, prefix uint64) OrderedNodeData {
+	return OrderedNodeData{
+		SortOrder: prefix,
+		NodeData: NodeData{
+			Key:   leaf.key,
+			Value: leaf.value,
+		},
+	}
+}
+
+// GetChunkHashes returns all the "checksum" hashes for
+// the chunks that will be sent
+func GetChunkHashes(tree *Tree, depth uint) ([][]byte, uint) {
+	// Account for unbalanced trees by capping the depth.
+	maxDepth := uint(tree.root.height / 2)
+	if depth > maxDepth {
+		depth = maxDepth
+	}
+
+	nodes := getNodes(tree, depth)
+	res := make([][]byte, len(nodes))
+	for i, n := range nodes {
+		res[i] = n.hash
+	}
+	return res, depth
+}
+
+// getNodes returns an array of nodes at the given depth
+func getNodes(tree *Tree, depth uint) []*Node {
+	nodes := make([]*Node, 0, 1<<depth)
+	tree.root.traverseDepth(tree, depth, func(node *Node) {
+		nodes = append(nodes, node)
+	})
+	return nodes
+}
+
+// call cb for every node exactly depth levels below it
+// depth first search to return in tree ordering
+func (node *Node) traverseDepth(t *Tree, depth uint, cb func(*Node)) {
+	// base case
+	if depth == 0 {
+		cb(node)
+		return
+	}
+
+	// otherwise, decend one more level
+	node.getLeftNode(t).traverseDepth(t, depth-1, cb)
+	node.getRightNode(t).traverseDepth(t, depth-1, cb)
+}
+
+// position to key can calculte the appropriate sort order
+// for the count-th node at a given depth, assuming a full
+// tree above this height.
+func positionToKey(depth, count uint) (key uint64) {
+	for d := depth; d > 0; d-- {
+		// lowest digit of count * 2^(d-1)
+		key += uint64((count & 1) << (d - 1))
+		count = count >> 1
+	}
+	return
+}
+
+// GetChunk finds the count-th subtree at depth and
+// generates a Chunk for that data
+func GetChunk(tree *Tree, depth, count uint) Chunk {
+	node := getNodes(tree, depth)[count]
+	prefix := positionToKey(depth, count)
+	return getChunk(tree, node, prefix, depth)
+}
+
+// getChunk takes a node and serializes all nodes below it
+//
+// As it is part of a larger tree, prefix defines the path
+// up to this point, and depth the current depth
+// (which defines where we add to the prefix)
+//
+// TODO: make this more efficient, *Chunk as arg???
+func getChunk(t *Tree, node *Node, prefix uint64, depth uint) Chunk {
+	if node.isLeaf() {
+		return Chunk{NewOrderedNode(node, prefix)}
+	}
+	res := make(Chunk, 0, node.size)
+	if node.leftNode != nil {
```

Comment:
shouldn't this be `getLeftNode` ? and same for `rightNode` below ?

Reply:
Please read the TODOs on the top.
All these algorithms assume db=nil. For quick development, to convince you all that this actually worked. They need to be modified to support a persistent versioned tree as the normal case.

getLeftNode() panics if there is no link and no backing db.